python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -471,43 +475,172 @@ def build_run_requests(
     asset_graph: AssetGraph,
     run_tags: Optional[Mapping[str, str]],
 ) -> Sequence[RunRequest]:
-    assets_to_reconcile_by_partitions_def_partition_key: Mapping[
-        Tuple[Optional[PartitionsDefinition], Optional[str]], Set[AssetKey]
-    ] = defaultdict(set)
+    backfill_policies = [
+        asset_graph.get_backfill_policy(asset_key)
+        for asset_key in set([asset_partition.asset_key for asset_partition in asset_partitions])
+    ]
+    all_assets_have_backfill_policies = backfill_policies and all(
+        backfill_policy is not None for backfill_policy in backfill_policies
+    )
+    run_requests = []
+    if all_assets_have_backfill_policies:
+        run_requests.extend(
+            _build_run_requests_with_backfill_policies(asset_partitions, asset_graph, run_tags)
+        )
+    else:
+        """
+        When any of the assets do not have backfill policies, we fall back to the default behavior of
+        backfilling them partition by partition.
+        """
+        assets_to_reconcile_by_partitions_def_partition_key: Mapping[
+            Tuple[Optional[PartitionsDefinition], Optional[str]], Set[AssetKey]
+        ] = defaultdict(set)
+
+        for asset_partition in asset_partitions:
+            assets_to_reconcile_by_partitions_def_partition_key[
+                asset_graph.get_partitions_def(asset_partition.asset_key),
+                asset_partition.partition_key,
+            ].add(asset_partition.asset_key)
+
+        for (
+            partitions_def,
+            partition_key,
+        ), asset_keys in assets_to_reconcile_by_partitions_def_partition_key.items():
+            tags = {**(run_tags or {})}
+            if partition_key is not None:
+                if partitions_def is None:
+                    check.failed("Partition key provided for unpartitioned asset")
+                tags.update({**partitions_def.get_tags_for_partition_key(partition_key)})
+
+            for asset_keys_in_repo in asset_graph.split_asset_keys_by_repository(asset_keys):
+                run_requests.append(
+                    # Do not call run_request.with_resolved_tags_and_config as the partition key is
+                    # valid and there is no config.
+                    # Calling with_resolved_tags_and_config is costly in asset reconciliation as it
+                    # checks for valid partition keys.
+                    RunRequest(
+                        asset_selection=list(asset_keys_in_repo),
+                        partition_key=partition_key,
+                        tags=tags,
+                    )
+                )
 
-    for asset_partition in asset_partitions:
-        assets_to_reconcile_by_partitions_def_partition_key[
-            asset_graph.get_partitions_def(asset_partition.asset_key), asset_partition.partition_key
-        ].add(asset_partition.asset_key)
+    return run_requests
 
+
+def _build_run_requests_with_backfill_policies(
```

Comment:
If I'm understanding how this function works correctly, I think there could be a problem in some situations. E.g. imagine that:
- Asset B depends on asset A
- Both have a single-run backfill policy and the same PartitionsDefinition
- Both are backfilled

If I understand the behavior correctly, this function will return two `RunRequest`s: one for A and one for B.  Assuming there aren't any queues, these runs will be executed concurrently. However, we'd actually want to wait until the materializations for A completed to launch the materializations for B. And if the materializations for A failed, we'd want to avoid launching B.

Reply:
> I assumed the runs will be picked up and executed sequentially, but it sounds like that is not the case here.

Ah, gotcha. Right, this is not the case. The way that the backfill engine respects execution dependencies, in the case that it can't put the upstream and downstream in the same run, is by waiting to submit a RunRequest for the downstream until after the upstream has finished materializing.